## Exploratory Data Analysis (EDA)

This is an exploratory data analysis of the 'Global Terrorism Database' (GTD)<sup>1</sup>, maintained by the National Consortium for the 'Study of Terrorism and Responses to Terrorism (START) at the University of Maryland. <br />
The primary dataset of terrorist attacks from 1970 to 2020 (globalterrorismdb_0522dist.xlsx), consists of approximately 210K records. For the purpose of this analysis, and considering performance, only the data for 2020 would be used (globalterrorismdb_0522dist_2020.xlsx), approximately 8.5K records.

Note:
<ul>
    <li>the data is a single worksheet containing all the relevant data including respective primary keys.</li>
    <li>all code was written by myself unless otherwise stated</li>
</ul>

Use the pandas library

In [1]:
import pandas as pd

Read data from the xlsx file, specifying the first row as the header

In [2]:
df = pd.read_excel('globalterrorismdb_0522dist_2020.xlsx', header=0)

Set the display to show all columns, set the dataframe index to the database id, 'eventid', and use the 'head' function to view the first five rows of data

In [3]:
pd.set_option('display.max_columns', None)
df.set_index('eventid')
df.head()

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  202001010001   2020       1     1        NaN         0        NaT      141   
1  202001010002   2020       1     1        NaN         0        NaT      141   
2  202001010003   2020       1     1        NaN         0        NaT      200   
3  202001010005   2020       1     1        NaN         0        NaT        4   
4  202001010006   2020       1     1        NaN         0        NaT       19   

   country_txt  region                  region_txt provstate      city  \
0        Nepal       6                  South Asia   Bagmati  Jurethum   
1        Nepal       6                  South Asia   Karnali    Pipira   
2        Syria      10  Middle East & North Africa    Raqqah     Suluk   
3  Afghanistan       6                  South Asia     Farah     Farah   
4   Bangladesh       6                  South Asia     Dhaka     Dhaka   

    latitude  longitude  specificity  vicinity  \
0  27.368102  85.107405            1         0   
1  28.572710  81.641440            1         0   
2  36.595594  39.128351            1         0   
3  32.366905  62.113825            1         0   
4  23.791275  90.418202            1         0   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  The incident occurred in the Pul-e-Jadid Regi ...   
4  The incident occurred near the Institute of En...   

                                             summary  crit1  crit2  crit3  \
0  01/01/2020: Assailants set fire to an Ncell te...      1      1      1   
1  01/01/2020: An explosive device was discovered...      1      1      1   
2  01/01/2020: An explosives-laden vehicle detona...      1      1      1   
3  01/01/2020: An explosive device detonated targ...      1      1      1   
4  01/01/2020: Two explosive devices detonated ou...      1      1      1   

   doubtterr  alternative alternative_txt  multiple  success  suicide  \
0          0          NaN             NaN       0.0        1        0   
1          0          NaN             NaN       0.0        0        0   
2          0          NaN             NaN       0.0        1        0   
3          0          NaN             NaN       0.0        1        0   
4          0          NaN             NaN       0.0        1        0   

   attacktype1                 attacktype1_txt  attacktype2 attacktype2_txt  \
0            7  Facility/Infrastructure Attack          NaN             NaN   
1            3               Bombing/Explosion          NaN             NaN   
2            3               Bombing/Explosion          NaN             NaN   
3            3               Bombing/Explosion          NaN             NaN   
4            3               Bombing/Explosion          NaN             NaN   

   attacktype3 attacktype3_txt  targtype1                targtype1_txt  \
0          NaN             NaN         16            Telecommunication   
1          NaN             NaN         16            Telecommunication   
2          NaN             NaN         14  Private Citizens & Property   
3          NaN             NaN         14  Private Citizens & Property   
4          NaN             NaN          8      Educational Institution   

   targsubtype1                        targsubtype1_txt  \
0          90.0                     Telephone/Telegraph   
1          90.0                     Telephone/Telegraph   
2          67.0            Unnamed Civilian/Unspecified   
3          73.0                 Vehicles/Transportation   
4          49.0  School/University/Educational Building   

                                     corp1                   target1  natlty1  \
0                    Ncell Private Limited  Telecommunications Tower    141.0   
1                    Ncell Private Limited  Telecommunications Tower    141.0   
2                 

Check the shape of the data

In [4]:
df.shape

(8438, 135)

Check the names of the columns. Note: use 'values.tolist()' to display all 135 column names.

In [5]:
df.columns.values.tolist()

['eventid',
 'iyear',
 'imonth',
 'iday',
 'approxdate',
 'extended',
 'resolution',
 'country',
 'country_txt',
 'region',
 'region_txt',
 'provstate',
 'city',
 'latitude',
 'longitude',
 'specificity',
 'vicinity',
 'location',
 'summary',
 'crit1',
 'crit2',
 'crit3',
 'doubtterr',
 'alternative',
 'alternative_txt',
 'multiple',
 'success',
 'suicide',
 'attacktype1',
 'attacktype1_txt',
 'attacktype2',
 'attacktype2_txt',
 'attacktype3',
 'attacktype3_txt',
 'targtype1',
 'targtype1_txt',
 'targsubtype1',
 'targsubtype1_txt',
 'corp1',
 'target1',
 'natlty1',
 'natlty1_txt',
 'targtype2',
 'targtype2_txt',
 'targsubtype2',
 'targsubtype2_txt',
 'corp2',
 'target2',
 'natlty2',
 'natlty2_txt',
 'targtype3',
 'targtype3_txt',
 'targsubtype3',
 'targsubtype3_txt',
 'corp3',
 'target3',
 'natlty3',
 'natlty3_txt',
 'gname',
 'gsubname',
 'gname2',
 'gsubname2',
 'gname3',
 'gsubname3',
 'motive',
 'guncertain1',
 'guncertain2',
 'guncertain3',
 'individual',
 'nperps',
 'nperpcap',
 

Set the display to show all rows and use 'dtypes' to view the datatype of each column.

In [6]:
pd.set_option('display.max_rows', None)
df.dtypes

eventid                        int64
iyear                          int64
imonth                         int64
iday                           int64
approxdate                    object
extended                       int64
resolution            datetime64[ns]
country                        int64
country_txt                   object
region                         int64
region_txt                    object
provstate                     object
city                          object
latitude                     float64
longitude                    float64
specificity                    int64
vicinity                       int64
location                      object
summary                       object
crit1                          int64
crit2                          int64
crit3                          int64
doubtterr                      int64
alternative                  float64
alternative_txt               object
multiple                     float64
success                        int64
s

Almost all the variables are either integers, floats or categorical.

Use 'nunique' to count the number of unique values for each column.

In [7]:
df.nunique()

eventid               8438
iyear                    1
imonth                  12
iday                    31
approxdate             354
extended                 2
resolution             136
country                101
country_txt            101
region                  12
region_txt              12
provstate              626
city                  4912
latitude              5947
longitude             5937
specificity              5
vicinity                 2
location              2676
summary               8342
crit1                    2
crit2                    2
crit3                    2
doubtterr                2
alternative              5
alternative_txt          5
multiple                 2
success                  2
suicide                  2
attacktype1              9
attacktype1_txt          9
attacktype2              5
attacktype2_txt          5
attacktype3              2
attacktype3_txt          2
targtype1               22
targtype1_txt           22
targsubtype1           107
t

Since there are 8438 rows of data, and there are the same amount of unique 'eventid', there is no need to remove any row of data based on 'eventid'.
As per the 'GTD Cookbook'<sup>2</sup>, the official documentation for the GTD, the variable 'eventid' is structured in the following format
<ul>
    <li>First 8 numbers - date recorded "yyyymmdd"</li>
    <li>Last 4 numbers - sequential case number for the given day (0001, 0002, etc)</li>
</ul>
Based on this information, the columns 'iyear', 'imonth' and 'iday' can be removed from the dataframe as they can be calculated from 'eventid' if required.

In [8]:
df.drop(['iyear', 'imonth', 'iday'], axis=1, inplace=True)
df.head()

eventid approxdate  extended resolution  country  country_txt  region  \
0  202001010001        NaN         0        NaT      141        Nepal       6   
1  202001010002        NaN         0        NaT      141        Nepal       6   
2  202001010003        NaN         0        NaT      200        Syria      10   
3  202001010005        NaN         0        NaT        4  Afghanistan       6   
4  202001010006        NaN         0        NaT       19   Bangladesh       6   

                   region_txt provstate      city   latitude  longitude  \
0                  South Asia   Bagmati  Jurethum  27.368102  85.107405   
1                  South Asia   Karnali    Pipira  28.572710  81.641440   
2  Middle East & North Africa    Raqqah     Suluk  36.595594  39.128351   
3                  South Asia     Farah     Farah  32.366905  62.113825   
4                  South Asia     Dhaka     Dhaka  23.791275  90.418202   

   specificity  vicinity                                           location  \
0            1         0                                                NaN   
1            1         0                                                NaN   
2            1         0                                                NaN   
3            1         0  The incident occurred in the Pul-e-Jadid Regi ...   
4            1         0  The incident occurred near the Institute of En...   

                                             summary  crit1  crit2  crit3  \
0  01/01/2020: Assailants set fire to an Ncell te...      1      1      1   
1  01/01/2020: An explosive device was discovered...      1      1      1   
2  01/01/2020: An explosives-laden vehicle detona...      1      1      1   
3  01/01/2020: An explosive device detonated targ...      1      1      1   
4  01/01/2020: Two explosive devices detonated ou...      1      1      1   

   doubtterr  alternative alternative_txt  multiple  success  suicide  \
0          0          NaN             NaN       0.0        1        0   
1          0          NaN             NaN       0.0        0        0   
2          0          NaN             NaN       0.0        1        0   
3          0          NaN             NaN       0.0        1        0   
4          0          NaN             NaN       0.0        1        0   

   attacktype1                 attacktype1_txt  attacktype2 attacktype2_txt  \
0            7  Facility/Infrastructure Attack          NaN             NaN   
1            3               Bombing/Explosion          NaN             NaN   
2            3               Bombing/Explosion          NaN             NaN   
3            3               Bombing/Explosion          NaN             NaN   
4            3               Bombing/Explosion          NaN             NaN   

   attacktype3 attacktype3_txt  targtype1                targtype1_txt  \
0          NaN             NaN         16            Telecommunication   
1          NaN             NaN         16            Telecommunication   
2          NaN             NaN         14  Private Citizens & Property   
3          NaN             NaN         14  Private Citizens & Property   
4          NaN             NaN          8      Educational Institution   

   targsubtype1                        targsubtype1_txt  \
0          90.0                     Telephone/Telegraph   
1          90.0                     Telephone/Telegraph   
2          67.0            Unnamed Civilian/Unspecified   
3          73.0                 Vehicles/Transportation   
4          49.0  School/University/Educational Building   

                                     corp1                   target1  natlty1  \
0                    Ncell Private Limited  Telecommunications Tower    141.0   
1                    Ncell Private Limited  Telecommunications Tower    141.0   
2                           Not Applicable                 Civilians    200.0   
3                           Not Applicable          Civilian Vehicle      4.0   
4  Institute 

For the purpose of the EDA, we would also drop 'approxdate', 'extended', and 'resolution'

In [9]:
df.drop(['approxdate', 'extended', 'resolution'], axis=1, inplace=True)
df.head()

eventid  country  country_txt  region                  region_txt  \
0  202001010001      141        Nepal       6                  South Asia   
1  202001010002      141        Nepal       6                  South Asia   
2  202001010003      200        Syria      10  Middle East & North Africa   
3  202001010005        4  Afghanistan       6                  South Asia   
4  202001010006       19   Bangladesh       6                  South Asia   

  provstate      city   latitude  longitude  specificity  vicinity  \
0   Bagmati  Jurethum  27.368102  85.107405            1         0   
1   Karnali    Pipira  28.572710  81.641440            1         0   
2    Raqqah     Suluk  36.595594  39.128351            1         0   
3     Farah     Farah  32.366905  62.113825            1         0   
4     Dhaka     Dhaka  23.791275  90.418202            1         0   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  The incident occurred in the Pul-e-Jadid Regi ...   
4  The incident occurred near the Institute of En...   

                                             summary  crit1  crit2  crit3  \
0  01/01/2020: Assailants set fire to an Ncell te...      1      1      1   
1  01/01/2020: An explosive device was discovered...      1      1      1   
2  01/01/2020: An explosives-laden vehicle detona...      1      1      1   
3  01/01/2020: An explosive device detonated targ...      1      1      1   
4  01/01/2020: Two explosive devices detonated ou...      1      1      1   

   doubtterr  alternative alternative_txt  multiple  success  suicide  \
0          0          NaN             NaN       0.0        1        0   
1          0          NaN             NaN       0.0        0        0   
2          0          NaN             NaN       0.0        1        0   
3          0          NaN             NaN       0.0        1        0   
4          0          NaN             NaN       0.0        1        0   

   attacktype1                 attacktype1_txt  attacktype2 attacktype2_txt  \
0            7  Facility/Infrastructure Attack          NaN             NaN   
1            3               Bombing/Explosion          NaN             NaN   
2            3               Bombing/Explosion          NaN             NaN   
3            3               Bombing/Explosion          NaN             NaN   
4            3               Bombing/Explosion          NaN             NaN   

   attacktype3 attacktype3_txt  targtype1                targtype1_txt  \
0          NaN             NaN         16            Telecommunication   
1          NaN             NaN         16            Telecommunication   
2          NaN             NaN         14  Private Citizens & Property   
3          NaN             NaN         14  Private Citizens & Property   
4          NaN             NaN          8      Educational Institution   

   targsubtype1                        targsubtype1_txt  \
0          90.0                     Telephone/Telegraph   
1          90.0                     Telephone/Telegraph   
2          67.0            Unnamed Civilian/Unspecified   
3          73.0                 Vehicles/Transportation   
4          49.0  School/University/Educational Building   

                                     corp1                   target1  natlty1  \
0                    Ncell Private Limited  Telecommunications Tower    141.0   
1                    Ncell Private Limited  Telecommunications Tower    141.0   
2                           Not Applicable                 Civilians    200.0   
3                           Not Applicable          Civilian Vehicle      4.0   
4  Institute of Engineers Bangladesh (IEB)               Institution     19.0   

   natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt corp2  \
0        Nepal        NaN           NaN    

Review the number of unique values for each column

In [10]:
df.nunique()

eventid               8438
country                101
country_txt            101
region                  12
region_txt              12
provstate              626
city                  4912
latitude              5947
longitude             5937
specificity              5
vicinity                 2
location              2676
summary               8342
crit1                    2
crit2                    2
crit3                    2
doubtterr                2
alternative              5
alternative_txt          5
multiple                 2
success                  2
suicide                  2
attacktype1              9
attacktype1_txt          9
attacktype2              5
attacktype2_txt          5
attacktype3              2
attacktype3_txt          2
targtype1               22
targtype1_txt           22
targsubtype1           107
targsubtype1_txt       107
corp1                 1894
target1               2606
natlty1                110
natlty1_txt            110
targtype2               18
t

Observe the following:<br />
<ul>
    <li>101 unique 'country' and 'country_txt'</li>
    <li>12 unique 'region' and 'region_txt'</li>
</ul>
This is a 2NF violation - 'country_txt' is dependent on 'country' and 'region_txt' is dependent on 'region', and not the primary key 'eventid'.<br />
This can be resolved by creating individual database tables for country and region.

Start with the 'region' and 'region_txt' columns as 'country' is dependent on 'region' as per the GTD Cookbook. <br />
Rename column 'region' to 'region_id', which would be used as the primary key of the 'region' database table.

In [11]:
df.rename(columns={"region": "region_id"}, inplace=True)
df.head()

eventid  country  country_txt  region_id                  region_txt  \
0  202001010001      141        Nepal          6                  South Asia   
1  202001010002      141        Nepal          6                  South Asia   
2  202001010003      200        Syria         10  Middle East & North Africa   
3  202001010005        4  Afghanistan          6                  South Asia   
4  202001010006       19   Bangladesh          6                  South Asia   

  provstate      city   latitude  longitude  specificity  vicinity  \
0   Bagmati  Jurethum  27.368102  85.107405            1         0   
1   Karnali    Pipira  28.572710  81.641440            1         0   
2    Raqqah     Suluk  36.595594  39.128351            1         0   
3     Farah     Farah  32.366905  62.113825            1         0   
4     Dhaka     Dhaka  23.791275  90.418202            1         0   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  The incident occurred in the Pul-e-Jadid Regi ...   
4  The incident occurred near the Institute of En...   

                                             summary  crit1  crit2  crit3  \
0  01/01/2020: Assailants set fire to an Ncell te...      1      1      1   
1  01/01/2020: An explosive device was discovered...      1      1      1   
2  01/01/2020: An explosives-laden vehicle detona...      1      1      1   
3  01/01/2020: An explosive device detonated targ...      1      1      1   
4  01/01/2020: Two explosive devices detonated ou...      1      1      1   

   doubtterr  alternative alternative_txt  multiple  success  suicide  \
0          0          NaN             NaN       0.0        1        0   
1          0          NaN             NaN       0.0        0        0   
2          0          NaN             NaN       0.0        1        0   
3          0          NaN             NaN       0.0        1        0   
4          0          NaN             NaN       0.0        1        0   

   attacktype1                 attacktype1_txt  attacktype2 attacktype2_txt  \
0            7  Facility/Infrastructure Attack          NaN             NaN   
1            3               Bombing/Explosion          NaN             NaN   
2            3               Bombing/Explosion          NaN             NaN   
3            3               Bombing/Explosion          NaN             NaN   
4            3               Bombing/Explosion          NaN             NaN   

   attacktype3 attacktype3_txt  targtype1                targtype1_txt  \
0          NaN             NaN         16            Telecommunication   
1          NaN             NaN         16            Telecommunication   
2          NaN             NaN         14  Private Citizens & Property   
3          NaN             NaN         14  Private Citizens & Property   
4          NaN             NaN          8      Educational Institution   

   targsubtype1                        targsubtype1_txt  \
0          90.0                     Telephone/Telegraph   
1          90.0                     Telephone/Telegraph   
2          67.0            Unnamed Civilian/Unspecified   
3          73.0                 Vehicles/Transportation   
4          49.0  School/University/Educational Building   

                                     corp1                   target1  natlty1  \
0                    Ncell Private Limited  Telecommunications Tower    141.0   
1                    Ncell Private Limited  Telecommunications Tower    141.0   
2                           Not Applicable                 Civilians    200.0   
3                           Not Applicable          Civilian Vehicle      4.0   
4  Institute of Engineers Bangladesh (IEB)               Institution     19.0   

   natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt corp2  \
0        Nepal        NaN

Create the 'region' dataframe of unique regions.

In [12]:
df_region = df[["region_id", "region_txt"]].copy().drop_duplicates()
df_region.head()

region_id                  region_txt
0           6                  South Asia
2          10  Middle East & North Africa
11         11          Sub-Saharan Africa
17          8              Western Europe
32          5              Southeast Asia

Change the 'region' dataframe index to 'region_id', sort by index and export to csv.

In [13]:
df_region.set_index('region_id').sort_index().to_csv('./csv/region_data.csv')

Drop the column 'region_txt' as it can be referenced by 'region_id'

In [14]:
df.drop('region_txt', axis=1, inplace=True)
df.head()

eventid  country  country_txt  region_id provstate      city  \
0  202001010001      141        Nepal          6   Bagmati  Jurethum   
1  202001010002      141        Nepal          6   Karnali    Pipira   
2  202001010003      200        Syria         10    Raqqah     Suluk   
3  202001010005        4  Afghanistan          6     Farah     Farah   
4  202001010006       19   Bangladesh          6     Dhaka     Dhaka   

    latitude  longitude  specificity  vicinity  \
0  27.368102  85.107405            1         0   
1  28.572710  81.641440            1         0   
2  36.595594  39.128351            1         0   
3  32.366905  62.113825            1         0   
4  23.791275  90.418202            1         0   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  The incident occurred in the Pul-e-Jadid Regi ...   
4  The incident occurred near the Institute of En...   

                                             summary  crit1  crit2  crit3  \
0  01/01/2020: Assailants set fire to an Ncell te...      1      1      1   
1  01/01/2020: An explosive device was discovered...      1      1      1   
2  01/01/2020: An explosives-laden vehicle detona...      1      1      1   
3  01/01/2020: An explosive device detonated targ...      1      1      1   
4  01/01/2020: Two explosive devices detonated ou...      1      1      1   

   doubtterr  alternative alternative_txt  multiple  success  suicide  \
0          0          NaN             NaN       0.0        1        0   
1          0          NaN             NaN       0.0        0        0   
2          0          NaN             NaN       0.0        1        0   
3          0          NaN             NaN       0.0        1        0   
4          0          NaN             NaN       0.0        1        0   

   attacktype1                 attacktype1_txt  attacktype2 attacktype2_txt  \
0            7  Facility/Infrastructure Attack          NaN             NaN   
1            3               Bombing/Explosion          NaN             NaN   
2            3               Bombing/Explosion          NaN             NaN   
3            3               Bombing/Explosion          NaN             NaN   
4            3               Bombing/Explosion          NaN             NaN   

   attacktype3 attacktype3_txt  targtype1                targtype1_txt  \
0          NaN             NaN         16            Telecommunication   
1          NaN             NaN         16            Telecommunication   
2          NaN             NaN         14  Private Citizens & Property   
3          NaN             NaN         14  Private Citizens & Property   
4          NaN             NaN          8      Educational Institution   

   targsubtype1                        targsubtype1_txt  \
0          90.0                     Telephone/Telegraph   
1          90.0                     Telephone/Telegraph   
2          67.0            Unnamed Civilian/Unspecified   
3          73.0                 Vehicles/Transportation   
4          49.0  School/University/Educational Building   

                                     corp1                   target1  natlty1  \
0                    Ncell Private Limited  Telecommunications Tower    141.0   
1                    Ncell Private Limited  Telecommunications Tower    141.0   
2                           Not Applicable                 Civilians    200.0   
3                           Not Applicable          Civilian Vehicle      4.0   
4  Institute of Engineers Bangladesh (IEB)               Institution     19.0   

   natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt corp2  \
0        Nepal        NaN           NaN           NaN              NaN   NaN   
1        Nepal        NaN           NaN           NaN              NaN   NaN   
2        Syria        NaN        

Rename column 'country' to 'country_id', which would be used as the primary key of the 'country' database table.

In [15]:
df.rename(columns={"country": "country_id"}, inplace=True)
df.head()

eventid  country_id  country_txt  region_id provstate      city  \
0  202001010001         141        Nepal          6   Bagmati  Jurethum   
1  202001010002         141        Nepal          6   Karnali    Pipira   
2  202001010003         200        Syria         10    Raqqah     Suluk   
3  202001010005           4  Afghanistan          6     Farah     Farah   
4  202001010006          19   Bangladesh          6     Dhaka     Dhaka   

    latitude  longitude  specificity  vicinity  \
0  27.368102  85.107405            1         0   
1  28.572710  81.641440            1         0   
2  36.595594  39.128351            1         0   
3  32.366905  62.113825            1         0   
4  23.791275  90.418202            1         0   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  The incident occurred in the Pul-e-Jadid Regi ...   
4  The incident occurred near the Institute of En...   

                                             summary  crit1  crit2  crit3  \
0  01/01/2020: Assailants set fire to an Ncell te...      1      1      1   
1  01/01/2020: An explosive device was discovered...      1      1      1   
2  01/01/2020: An explosives-laden vehicle detona...      1      1      1   
3  01/01/2020: An explosive device detonated targ...      1      1      1   
4  01/01/2020: Two explosive devices detonated ou...      1      1      1   

   doubtterr  alternative alternative_txt  multiple  success  suicide  \
0          0          NaN             NaN       0.0        1        0   
1          0          NaN             NaN       0.0        0        0   
2          0          NaN             NaN       0.0        1        0   
3          0          NaN             NaN       0.0        1        0   
4          0          NaN             NaN       0.0        1        0   

   attacktype1                 attacktype1_txt  attacktype2 attacktype2_txt  \
0            7  Facility/Infrastructure Attack          NaN             NaN   
1            3               Bombing/Explosion          NaN             NaN   
2            3               Bombing/Explosion          NaN             NaN   
3            3               Bombing/Explosion          NaN             NaN   
4            3               Bombing/Explosion          NaN             NaN   

   attacktype3 attacktype3_txt  targtype1                targtype1_txt  \
0          NaN             NaN         16            Telecommunication   
1          NaN             NaN         16            Telecommunication   
2          NaN             NaN         14  Private Citizens & Property   
3          NaN             NaN         14  Private Citizens & Property   
4          NaN             NaN          8      Educational Institution   

   targsubtype1                        targsubtype1_txt  \
0          90.0                     Telephone/Telegraph   
1          90.0                     Telephone/Telegraph   
2          67.0            Unnamed Civilian/Unspecified   
3          73.0                 Vehicles/Transportation   
4          49.0  School/University/Educational Building   

                                     corp1                   target1  natlty1  \
0                    Ncell Private Limited  Telecommunications Tower    141.0   
1                    Ncell Private Limited  Telecommunications Tower    141.0   
2                           Not Applicable                 Civilians    200.0   
3                           Not Applicable          Civilian Vehicle      4.0   
4  Institute of Engineers Bangladesh (IEB)               Institution     19.0   

   natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt corp2  \
0        Nepal        NaN           NaN           NaN              NaN   NaN   
1        Nepal        NaN           NaN           NaN              NaN   NaN   
2        Syria 

Create the 'country' dataframe of unique values. <br />
'region_id' is required as it would be used as a 'foreign key' in the 'country' SQL database.

In [16]:
df_country = df[['country_id', 'country_txt', 'region_id']].copy().drop_duplicates()
df_country.head()

country_id  country_txt  region_id
0         141        Nepal          6
2         200        Syria         10
3           4  Afghanistan          6
4          19   Bangladesh          6
8         153     Pakistan          6

Change the 'country' dataframe index to 'country_id', sort by index and export to csv.

In [17]:
df_country.set_index('country_id').sort_index().to_csv('./csv/country_data.csv')

Drop the column 'country_txt' as it can be referenced by 'country_id'

In [18]:
df.drop('country_txt', axis=1, inplace=True)
df.head()

eventid  country_id  region_id provstate      city   latitude  \
0  202001010001         141          6   Bagmati  Jurethum  27.368102   
1  202001010002         141          6   Karnali    Pipira  28.572710   
2  202001010003         200         10    Raqqah     Suluk  36.595594   
3  202001010005           4          6     Farah     Farah  32.366905   
4  202001010006          19          6     Dhaka     Dhaka  23.791275   

   longitude  specificity  vicinity  \
0  85.107405            1         0   
1  81.641440            1         0   
2  39.128351            1         0   
3  62.113825            1         0   
4  90.418202            1         0   

                                            location  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  The incident occurred in the Pul-e-Jadid Regi ...   
4  The incident occurred near the Institute of En...   

                                             summary  crit1  crit2  crit3  \
0  01/01/2020: Assailants set fire to an Ncell te...      1      1      1   
1  01/01/2020: An explosive device was discovered...      1      1      1   
2  01/01/2020: An explosives-laden vehicle detona...      1      1      1   
3  01/01/2020: An explosive device detonated targ...      1      1      1   
4  01/01/2020: Two explosive devices detonated ou...      1      1      1   

   doubtterr  alternative alternative_txt  multiple  success  suicide  \
0          0          NaN             NaN       0.0        1        0   
1          0          NaN             NaN       0.0        0        0   
2          0          NaN             NaN       0.0        1        0   
3          0          NaN             NaN       0.0        1        0   
4          0          NaN             NaN       0.0        1        0   

   attacktype1                 attacktype1_txt  attacktype2 attacktype2_txt  \
0            7  Facility/Infrastructure Attack          NaN             NaN   
1            3               Bombing/Explosion          NaN             NaN   
2            3               Bombing/Explosion          NaN             NaN   
3            3               Bombing/Explosion          NaN             NaN   
4            3               Bombing/Explosion          NaN             NaN   

   attacktype3 attacktype3_txt  targtype1                targtype1_txt  \
0          NaN             NaN         16            Telecommunication   
1          NaN             NaN         16            Telecommunication   
2          NaN             NaN         14  Private Citizens & Property   
3          NaN             NaN         14  Private Citizens & Property   
4          NaN             NaN          8      Educational Institution   

   targsubtype1                        targsubtype1_txt  \
0          90.0                     Telephone/Telegraph   
1          90.0                     Telephone/Telegraph   
2          67.0            Unnamed Civilian/Unspecified   
3          73.0                 Vehicles/Transportation   
4          49.0  School/University/Educational Building   

                                     corp1                   target1  natlty1  \
0                    Ncell Private Limited  Telecommunications Tower    141.0   
1                    Ncell Private Limited  Telecommunications Tower    141.0   
2                           Not Applicable                 Civilians    200.0   
3                           Not Applicable          Civilian Vehicle      4.0   
4  Institute of Engineers Bangladesh (IEB)               Institution     19.0   

   natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt corp2  \
0        Nepal        NaN           NaN           NaN              NaN   NaN   
1        Nepal        NaN           NaN           NaN              NaN   NaN   
2        Syria        NaN           NaN           NaN              NaN   NaN   
3  Afghanista

For the purpose of this EDA, data rows where either 'provstate' or 'city' have values 'unknown' are dropped. This would also elimintate rows of data with unknown 'latitude' and 'longitude', and would correspond to column 'specificity' with values '4' and '5', as noted in the 'GTD Cookbook', page 23-24. <br />
It's is to be noted that this would also eliminate 4 rows of data where 'city' has the value 'Uknown', clearly a typo.

In [19]:
index_unknown_provstate_city = df[(df['specificity'] == 4) | (df['specificity'] == 5)].index
df.drop(index_unknown_provstate_city, inplace=True)
df.shape

(8089, 127)

Drop 'latitude', 'longitude', 'specificity', 'vicinity', 'location' and 'summary' as the data in these columns are specific to the event.

In [20]:
df.drop(['latitude', 'longitude', 'specificity', 'vicinity', 'location', 'summary'], axis=1, inplace=True)
df.head()

eventid  country_id  region_id provstate      city  crit1  crit2  \
0  202001010001         141          6   Bagmati  Jurethum      1      1   
1  202001010002         141          6   Karnali    Pipira      1      1   
2  202001010003         200         10    Raqqah     Suluk      1      1   
3  202001010005           4          6     Farah     Farah      1      1   
4  202001010006          19          6     Dhaka     Dhaka      1      1   

   crit3  doubtterr  alternative alternative_txt  multiple  success  suicide  \
0      1          0          NaN             NaN       0.0        1        0   
1      1          0          NaN             NaN       0.0        0        0   
2      1          0          NaN             NaN       0.0        1        0   
3      1          0          NaN             NaN       0.0        1        0   
4      1          0          NaN             NaN       0.0        1        0   

   attacktype1                 attacktype1_txt  attacktype2 attacktype2_txt  \
0            7  Facility/Infrastructure Attack          NaN             NaN   
1            3               Bombing/Explosion          NaN             NaN   
2            3               Bombing/Explosion          NaN             NaN   
3            3               Bombing/Explosion          NaN             NaN   
4            3               Bombing/Explosion          NaN             NaN   

   attacktype3 attacktype3_txt  targtype1                targtype1_txt  \
0          NaN             NaN         16            Telecommunication   
1          NaN             NaN         16            Telecommunication   
2          NaN             NaN         14  Private Citizens & Property   
3          NaN             NaN         14  Private Citizens & Property   
4          NaN             NaN          8      Educational Institution   

   targsubtype1                        targsubtype1_txt  \
0          90.0                     Telephone/Telegraph   
1          90.0                     Telephone/Telegraph   
2          67.0            Unnamed Civilian/Unspecified   
3          73.0                 Vehicles/Transportation   
4          49.0  School/University/Educational Building   

                                     corp1                   target1  natlty1  \
0                    Ncell Private Limited  Telecommunications Tower    141.0   
1                    Ncell Private Limited  Telecommunications Tower    141.0   
2                           Not Applicable                 Civilians    200.0   
3                           Not Applicable          Civilian Vehicle      4.0   
4  Institute of Engineers Bangladesh (IEB)               Institution     19.0   

   natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt corp2  \
0        Nepal        NaN           NaN           NaN              NaN   NaN   
1        Nepal        NaN           NaN           NaN              NaN   NaN   
2        Syria        NaN           NaN           NaN              NaN   NaN   
3  Afghanistan        NaN           NaN           NaN              NaN   NaN   
4   Bangladesh        NaN           NaN           NaN              NaN   NaN   

  target2  natlty2 natlty2_txt  targtype3 targtype3_txt  targsubtype3  \
0     NaN      NaN         NaN        NaN           NaN           NaN   
1     NaN      NaN         NaN        NaN           NaN           NaN   
2     NaN      NaN         NaN        NaN           NaN           NaN   
3     NaN      NaN         NaN        NaN           NaN           NaN   
4     NaN      NaN         NaN        NaN           NaN           NaN   

  targsubtype3_txt corp3 target3  natlty3 natlty3_txt  \
0              NaN   NaN     NaN      NaN         NaN   
1              NaN   NaN     NaN      NaN         NaN   
2              NaN   NaN     NaN      NaN         NaN   
3              NaN   NaN     NaN      NaN         NaN   
4              NaN   NaN     NaN      NaN         NaN   

                                             

Re-run dataframe structure commands

In [21]:
df.dtypes

eventid                 int64
country_id              int64
region_id               int64
provstate              object
city                   object
crit1                   int64
crit2                   int64
crit3                   int64
doubtterr               int64
alternative           float64
alternative_txt        object
multiple              float64
success                 int64
suicide                 int64
attacktype1             int64
attacktype1_txt        object
attacktype2           float64
attacktype2_txt        object
attacktype3           float64
attacktype3_txt        object
targtype1               int64
targtype1_txt          object
targsubtype1          float64
targsubtype1_txt       object
corp1                  object
target1                object
natlty1               float64
natlty1_txt            object
targtype2             float64
targtype2_txt          object
targsubtype2          float64
targsubtype2_txt       object
corp2                  object
target2   

In [22]:
df.nunique()

eventid               8089
country_id             101
region_id               12
provstate              617
city                  4910
crit1                    2
crit2                    2
crit3                    2
doubtterr                2
alternative              5
alternative_txt          5
multiple                 2
success                  2
suicide                  2
attacktype1              9
attacktype1_txt          9
attacktype2              5
attacktype2_txt          5
attacktype3              2
attacktype3_txt          2
targtype1               22
targtype1_txt           22
targsubtype1           107
targsubtype1_txt       107
corp1                 1861
target1               2527
natlty1                110
natlty1_txt            110
targtype2               18
targtype2_txt           18
targsubtype2            75
targsubtype2_txt        75
corp2                  217
target2                342
natlty2                 61
natlty2_txt             61
targtype3               12
t

There are several columns with two unique values. These are typical '1=Yes, 0=No' columns. After verifying with the 'GTD Cookbook', convert these to the boolean datatype.

Convert 'crit1', 'crit2', 'crit3', 'doubtterr', 'multiple', 'success', and 'suicide' to boolean datatype.

In [23]:
# list to convert multiple columns
bool_list = ['crit1', 'crit2', 'crit3', 'doubtterr', 'multiple', 'success', 'suicide']

for col in bool_list:
    df[col] = df[col].astype('bool')

In [24]:
df.dtypes

eventid                 int64
country_id              int64
region_id               int64
provstate              object
city                   object
crit1                    bool
crit2                    bool
crit3                    bool
doubtterr                bool
alternative           float64
alternative_txt        object
multiple                 bool
success                  bool
suicide                  bool
attacktype1             int64
attacktype1_txt        object
attacktype2           float64
attacktype2_txt        object
attacktype3           float64
attacktype3_txt        object
targtype1               int64
targtype1_txt          object
targsubtype1          float64
targsubtype1_txt       object
corp1                  object
target1                object
natlty1               float64
natlty1_txt            object
targtype2             float64
targtype2_txt          object
targsubtype2          float64
targsubtype2_txt       object
corp2                  object
target2   

As per the 'GTD Cookbook', page 17, 'alternative' and 'alternative_txt' are only used when 'doubtterr' = '1' or 'Yes', e.i. null values are permitted, hence python is indicating the 'float64' data type for column 'alternative'.

Rename column 'alternative' to 'alternative_id', which would be the primary key of the 'alternative_dsn' database table.

In [25]:
df.rename(columns={"alternative": "alternative_id"}, inplace=True)
df.head()

eventid  country_id  region_id provstate      city  crit1  crit2  \
0  202001010001         141          6   Bagmati  Jurethum   True   True   
1  202001010002         141          6   Karnali    Pipira   True   True   
2  202001010003         200         10    Raqqah     Suluk   True   True   
3  202001010005           4          6     Farah     Farah   True   True   
4  202001010006          19          6     Dhaka     Dhaka   True   True   

   crit3  doubtterr  alternative_id alternative_txt  multiple  success  \
0   True      False             NaN             NaN     False     True   
1   True      False             NaN             NaN     False    False   
2   True      False             NaN             NaN     False     True   
3   True      False             NaN             NaN     False     True   
4   True      False             NaN             NaN     False     True   

   suicide  attacktype1                 attacktype1_txt  attacktype2  \
0    False            7  Facility/Infrastructure Attack          NaN   
1    False            3               Bombing/Explosion          NaN   
2    False            3               Bombing/Explosion          NaN   
3    False            3               Bombing/Explosion          NaN   
4    False            3               Bombing/Explosion          NaN   

  attacktype2_txt  attacktype3 attacktype3_txt  targtype1  \
0             NaN          NaN             NaN         16   
1             NaN          NaN             NaN         16   
2             NaN          NaN             NaN         14   
3             NaN          NaN             NaN         14   
4             NaN          NaN             NaN          8   

                 targtype1_txt  targsubtype1  \
0            Telecommunication          90.0   
1            Telecommunication          90.0   
2  Private Citizens & Property          67.0   
3  Private Citizens & Property          73.0   
4      Educational Institution          49.0   

                         targsubtype1_txt  \
0                     Telephone/Telegraph   
1                     Telephone/Telegraph   
2            Unnamed Civilian/Unspecified   
3                 Vehicles/Transportation   
4  School/University/Educational Building   

                                     corp1                   target1  natlty1  \
0                    Ncell Private Limited  Telecommunications Tower    141.0   
1                    Ncell Private Limited  Telecommunications Tower    141.0   
2                           Not Applicable                 Civilians    200.0   
3                           Not Applicable          Civilian Vehicle      4.0   
4  Institute of Engineers Bangladesh (IEB)               Institution     19.0   

   natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt corp2  \
0        Nepal        NaN           NaN           NaN              NaN   NaN   
1        Nepal        NaN           NaN           NaN              NaN   NaN   
2        Syria        NaN           NaN           NaN              NaN   NaN   
3  Afghanistan        NaN           NaN           NaN              NaN   NaN   
4   Bangladesh        NaN           NaN           NaN              NaN   NaN   

  target2  natlty2 natlty2_txt  targtype3 targtype3_txt  targsubtype3  \
0     NaN      NaN         NaN        NaN           NaN           NaN   
1     NaN      NaN         NaN        NaN           NaN           NaN   
2     NaN      NaN         NaN        NaN           NaN           NaN   
3     NaN      NaN         NaN        NaN           NaN           NaN   
4     NaN      NaN         NaN        NaN           NaN           NaN   

  targsubtype3_txt corp3 target3  natlty3 natlty3_txt  \
0              NaN   NaN     NaN      NaN         NaN   
1              NaN   NaN     NaN      NaN         NaN   
2              NaN   NaN     NaN      NaN         NaN   
3              NaN   NaN     NaN      NaN         NaN   
4              NaN   NaN     NaN      NaN         NaN 

Create the 'alternative_dsn' dataframe of unique values, dropping the 'alternative_id' that is Nan or null.

In [26]:
df_alternative_dsn = df[["alternative_id", "alternative_txt"]].copy().drop_duplicates().dropna()
df_alternative_dsn.head()

alternative_id             alternative_txt
6               1.0  Insurgency/Guerilla Action
17              2.0            Other Crime Type
71              4.0      Lack of Intentionality
122             5.0                State Actors
174             3.0  Intra/Inter-group Conflict

Change the data type of 'alternative_id' to int64.<br />
Change the 'alternative_dsn' dataframe index to 'alternative_id', sort by index and export to csv.

In [27]:
df_alternative_dsn['alternative_id'] = df_alternative_dsn['alternative_id'].astype('int64')
df_alternative_dsn.set_index('alternative_id').sort_index().to_csv('./csv/alternative_dsn_data.csv')

Drop the column 'alternative_txt' as it can be referenced by 'alternative_id'

In [28]:
df.drop('alternative_txt', axis=1, inplace=True)
df.head()

eventid  country_id  region_id provstate      city  crit1  crit2  \
0  202001010001         141          6   Bagmati  Jurethum   True   True   
1  202001010002         141          6   Karnali    Pipira   True   True   
2  202001010003         200         10    Raqqah     Suluk   True   True   
3  202001010005           4          6     Farah     Farah   True   True   
4  202001010006          19          6     Dhaka     Dhaka   True   True   

   crit3  doubtterr  alternative_id  multiple  success  suicide  attacktype1  \
0   True      False             NaN     False     True    False            7   
1   True      False             NaN     False    False    False            3   
2   True      False             NaN     False     True    False            3   
3   True      False             NaN     False     True    False            3   
4   True      False             NaN     False     True    False            3   

                  attacktype1_txt  attacktype2 attacktype2_txt  attacktype3  \
0  Facility/Infrastructure Attack          NaN             NaN          NaN   
1               Bombing/Explosion          NaN             NaN          NaN   
2               Bombing/Explosion          NaN             NaN          NaN   
3               Bombing/Explosion          NaN             NaN          NaN   
4               Bombing/Explosion          NaN             NaN          NaN   

  attacktype3_txt  targtype1                targtype1_txt  targsubtype1  \
0             NaN         16            Telecommunication          90.0   
1             NaN         16            Telecommunication          90.0   
2             NaN         14  Private Citizens & Property          67.0   
3             NaN         14  Private Citizens & Property          73.0   
4             NaN          8      Educational Institution          49.0   

                         targsubtype1_txt  \
0                     Telephone/Telegraph   
1                     Telephone/Telegraph   
2            Unnamed Civilian/Unspecified   
3                 Vehicles/Transportation   
4  School/University/Educational Building   

                                     corp1                   target1  natlty1  \
0                    Ncell Private Limited  Telecommunications Tower    141.0   
1                    Ncell Private Limited  Telecommunications Tower    141.0   
2                           Not Applicable                 Civilians    200.0   
3                           Not Applicable          Civilian Vehicle      4.0   
4  Institute of Engineers Bangladesh (IEB)               Institution     19.0   

   natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt corp2  \
0        Nepal        NaN           NaN           NaN              NaN   NaN   
1        Nepal        NaN           NaN           NaN              NaN   NaN   
2        Syria        NaN           NaN           NaN              NaN   NaN   
3  Afghanistan        NaN           NaN           NaN              NaN   NaN   
4   Bangladesh        NaN           NaN           NaN              NaN   NaN   

  target2  natlty2 natlty2_txt  targtype3 targtype3_txt  targsubtype3  \
0     NaN      NaN         NaN        NaN           NaN           NaN   
1     NaN      NaN         NaN        NaN           NaN           NaN   
2     NaN      NaN         NaN        NaN           NaN           NaN   
3     NaN      NaN         NaN        NaN           NaN           NaN   
4     NaN      NaN         NaN        NaN           NaN           NaN   

  targsubtype3_txt corp3 target3  natlty3 natlty3_txt  \
0              NaN   NaN     NaN      NaN         NaN   
1              NaN   NaN     NaN      NaN         NaN   
2              NaN   NaN     NaN      NaN         NaN   
3              NaN   NaN     NaN      NaN         NaN   
4              NaN   NaN     NaN      NaN         NaN   

                                               gname gsubname gname2  \
0  Communist Party of Nepal - Maoist (CPN-Maoist-..

There are three each of columns 'attacktype*' and 'attacktype*_txt'.<br />
Each event must have at least one attack type. The other fields are only used when there are multiple attack types used by the perpetrators. At the same time, this structure also limits the data to a maximum of three attack types.
<br />
This is a violation of:<br />
<ul>
    <li>1NF - repeating groups of data, which can be resolved by creating a 'bridging' table</li>
    <li>2NF - dependency on another field and not the primary key, for example, 'attacktype_txt' is dependent on 'attacktype'</li>
</ul>
Furthermore, each 'attacktype*' and 'attacktype*_txt' have the same number of unique values.<br />
For the SQL database, data extraction would be two fold:
<ol>
    <li>Use 'attacktype1' and 'attacktype1_txt' to create the data for the 'attack_type' database table</li>
    <li>Create a dataframe that would be the source of a SQL bridging table for the multiple attack types used in each terrorism event.</li>
</ol>
Note: The data for the bridging table cannot be extracted unless the data has been 'cleaned'.

Create the 'attack_type' dataframe of unique values.

In [29]:
df_attack_type = df[["attacktype1", "attacktype1_txt"]].copy().drop_duplicates()
df_attack_type.head()

attacktype1                 attacktype1_txt
0             7  Facility/Infrastructure Attack
1             3               Bombing/Explosion
5             9                         Unknown
9             1                   Assassination
12            2                   Armed Assault

Rename columns in the 'attack_type' dataframe.<br />
Change the dataframe index to 'attack_type_id', sort by index and export to csv.

In [30]:
df_attack_type.rename(columns={"attacktype1": "attack_type_id"}, inplace=True)
df_attack_type.rename(columns={"attacktype1_txt": "attack_type_txt"}, inplace=True)
df_attack_type.set_index('attack_type_id').sort_index().to_csv('./csv/attack_type_data.csv')

There are three each of columns 'targtype*', 'targtype*_txt', 'targsubtype*', 'targsubtype*_txt', 'natlty*' and 'natlty*_txt' - 'corp*' and 'target*' would not be used since it is specific to the event, and would be dropped.<br />
This is a violation of:<br />
<ul>
    <li>1NF - repeating groups of data, which can be resolved by creating a 'bridging' table.</li>
    <li>2NF - dependency on another field and not the primary key, for example, 'targtype*_txt' is dependent on 'targtype*', and not the primary key 'eventid'.</li>
</ul>
Furthermore, each respective pair of 'targtype*' and 'targtype*_txt', 'targsubtype*' and 'targsubtype*_txt', and 'natlty*' and 'natlty*_txt' have the same number of unique values.<br />
For the SQL database, data extraction would be two fold:
<ol>
    <li>Iterate over each of 'targtype*', 'targtype*_txt', 'targsubtype*', 'targsubtype*_txt', 'natlty1' and 'natlty1_txt', to create the data for the respective 'target_type', 'target_subtype' and 'nationality' database tables.</li>
    <li>Create a dataframe that would be the source of a SQL bridging table for the targets of the terrorism event.</li>
</ol>
Note: The data for the bridging table cannot be extracted unless the data has been 'cleaned'.

First, drop rows where 'targsubtype1' is Nan or null.<br />
This mostly correspond to the 'targtype_txt' of value 'Unknown'

In [31]:
df = df.dropna(subset=['targsubtype1'])
df.shape

(7765, 120)

Iterate over each row to eventually create the 'target_type', 'target_subtype' and 'nationality' dataframes

In [32]:
import math

# empty dictionary containers
target_types = {}
target_subtypes = {}
nationalities = {}

for row in df.itertuples():
    # key, value pair using targtype*, targtype*_txt
    target_types[int(row.targtype1)] = row.targtype1_txt
    # key, value pair using targsubtype*, targsubtype*_txt
    target_subtypes[int(row.targsubtype1)] = row.targsubtype1_txt
    # key, value pair using natlty*, natlty*_txt
    nationalities[int(row.natlty1)] = row.natlty1_txt

    # if not empty, update target_types
    if not math.isnan(row.targtype2):
        target_types[int(row.targtype2)] = row.targtype2_txt

    # if not empty, update target_subtypes
    if not math.isnan(row.targsubtype2):
        target_subtypes[int(row.targsubtype2)] = row.targsubtype2_txt

    # if not empty, update nationalities
    if not math.isnan(row.natlty2):
        nationalities[int(row.natlty2)] = row.natlty2_txt

    # if not empty, update target_types
    if not math.isnan(row.targtype3):
        target_types[int(row.targtype3)] = row.targtype3_txt

    # if not empty, update target_subtypes
    if not math.isnan(row.targsubtype3):
        target_subtypes[int(row.targsubtype3)] = row.targsubtype3_txt

    # if not empty, update nationalities
    if not math.isnan(row.natlty3):
        nationalities[int(row.natlty3)] = row.natlty3_txt

# create the target_type dataframe specifying the comumn names
df_target_type = pd.DataFrame(target_types.items(), columns=['target_type_id', 'target_type_txt'])
# create the target_subtype dataframe specifying the comumn names
df_target_subtype = pd.DataFrame(target_subtypes.items(), columns=['target_subtype_id', 'target_subtype_txt'])
# create the nationality dataframe specifying the comumn names
df_nationality = pd.DataFrame(nationalities.items(), columns=['nationality_id', 'nationality_txt'])

For each of the dataframes above, change the index to the '*_id', sort by index and export to csv.

In [33]:
# target_type
df_target_type.set_index('target_type_id').sort_index().to_csv('./csv/target_type_data.csv')
# target_subtype
df_target_subtype.set_index('target_subtype_id').sort_index().to_csv('./csv/target_subtype_data.csv')
# nationality
df_nationality.set_index('nationality_id').sort_index().to_csv('./csv/nationality_data.csv')

There are four each of columns 'weaptype*', 'weaptype*_txt', 'weapsubtype*', and 'weapsubtype*_txt'.<br />
This is a violation of:<br />
<ul>
    <li>1NF - repeating groups of data, which can be resolved by creating a 'bridging' table.</li>
    <li>2NF - dependency on another field and not the primary key, for example, 'weaptype*_txt' is dependent on 'weaptype*', and not the primary key 'eventid'.</li>
</ul>
Furthermore, each respective pair of 'weaptype*' and 'weaptype*_txt', and 'weapsubtype*' and 'weapsubtype*_txt' have the same number of unique values.<br />
For the SQL database, data extraction would be two fold:
<ol>
    <li>Iterate over each of 'weaptype*', 'weaptype*_txt', 'weapsubtype*', and 'weapsubtype*_txt', to create the data for the respective 'weapon_type', and 'weapon_subtype' database tables.</li>
    <li>Create a dataframe that would be the source of a SQL bridging table for the weapons of the terrorism event.</li>
</ol>
Note: The data for the bridging table cannot be extracted unless the data has been 'cleaned'.

First, drop rows where 'wepsubtype1' is NaN or null.
This mostly correspond to the weapgtype_txt' of value 'Unknown'

In [34]:
df = df.dropna(subset=['weapsubtype1'])
df.shape

(6107, 120)

Iterate over each row to eventually create the 'weapon_type', and 'weapon_subtype' dataframes

In [35]:
import math

# empty dictionary containers
weapon_types = {}
weapon_subtypes = {}

for row in df.itertuples():
    # key, value pair using weaptype*, weaptype*_txt
    weapon_types[int(row.weaptype1)] = row.weaptype1_txt
    # key, value pair using weapsubtype*, weapsubtype*_txt
    weapon_subtypes[int(row.weapsubtype1)] = row.weapsubtype1_txt

    # if not empty, update weapon_types
    if not math.isnan(row.weaptype2):
        weapon_types[int(row.weaptype2)] = row.weaptype2_txt

    # if not empty, update weapon_subtypes
    if not math.isnan(row.weapsubtype2):
        weapon_subtypes[int(row.weapsubtype2)] = row.weapsubtype2_txt

    # if not empty, update weapon_types
    if not math.isnan(row.weaptype3):
        weapon_types[int(row.weaptype3)] = row.weaptype3_txt

    # if not empty, update weapon_subtypes
    if not math.isnan(row.weapsubtype3):
        weapon_subtypes[int(row.weapsubtype3)] = row.weapsubtype3_txt

    # if not empty, update weapon_types
    if not math.isnan(row.weaptype4):
        weapon_types[int(row.weaptype4)] = row.weaptype4_txt

    # if not empty, update weapon_subtypes
    if not math.isnan(row.weapsubtype4):
        weapon_subtypes[int(row.weapsubtype4)] = row.weapsubtype4_txt

# create the weapon_type dataframe specifying the comumn names
df_weapon_type = pd.DataFrame(weapon_types.items(), columns=['weapon_type_id', 'weapon_type_txt'])
# create the weapon_subtype dataframe specifying the comumn names
df_weapon_subtype = pd.DataFrame(weapon_subtypes.items(), columns=['weapon_subtype_id', 'weapon_subtype_txt'])

For each of the dataframes above, change the index to the '*_id', sort by index and export to csv.

In [36]:
# weapon_type
df_weapon_type.set_index('weapon_type_id').sort_index().to_csv('./csv/weapon_type_data.csv')
# weapon_subtype
df_weapon_subtype.set_index('weapon_subtype_id').sort_index().to_csv('./csv/weapon_subtype_data.csv')

The only other columns that are of interest are 'nkill', 'nkillus', and 'nkillter'. All other columns of data are specific to the event or has sparse information.<br />
According to the 'GTD Cookbook', if the figure is not reported and is too vague to be of use, these fields would be blank.<br />
Only 'nkill' and 'nkillter' have blank values.

First, drop rows where 'nkill' is NaN or null.

In [37]:
df = df.dropna(subset=['nkill'])
df.shape

(5834, 120)

Drop rows where 'nkillter' is NaN or null

In [38]:
df = df.dropna(subset=['nkillter'])
df.shape

(5782, 120)

At this stage, the required data is clean and the data for the bridging tables can be generated.<br />
Iterate through the main dataframe, to create the list of eventids and the various respective attack_type_ids.<br />
This list is then used to create the 'event_attack_type' dataframe.

In [39]:
import math
# initialize an empty list
event_attack_types = []

for row in df.itertuples():
    # append eventid and attacktype1
    event_attack_types.append({'event_id': row.eventid, 'attack_type_id': int(row.attacktype1)})

    # if not empty, append eventid and attacktype2
    if not math.isnan(row.attacktype2):
        event_attack_types.append({'event_id': row.eventid, 'attack_type_id': int(row.attacktype2)})

    # if not empty, append eventid and attacktype3
    if not math.isnan(row.attacktype3):
        event_attack_types.append({'event_id': row.eventid, 'attack_type_id': int(row.attacktype3)})

df_event_attack_type = pd.DataFrame(event_attack_types)
df_event_attack_type.head()

event_id  attack_type_id
0  202001010001               7
1  202001010002               3
2  202001010003               3
3  202001010005               3
4  202001010006               3

Change the 'event_attack_type' dataframe index to 'event_id' and 'attack_type_id', and export to csv.<br />
Both 'event_id' and 'attack_type_id' would be used as primary keys in the SQL table.

In [40]:
df_event_attack_type.set_index(['event_id', 'attack_type_id']).to_csv('./csv/event_attack_type_data.csv')

Drop all 'attacktype*' and 'attacktype*_txt' columns.

In [41]:
df.drop(
    [
        'attacktype1', 'attacktype1_txt',
        'attacktype2', 'attacktype2_txt', 
        'attacktype3', 'attacktype3_txt'
    ], 
    axis=1, inplace=True
)

Re-run dataframe structure commands

In [42]:
df.dtypes

eventid                 int64
country_id              int64
region_id               int64
provstate              object
city                   object
crit1                    bool
crit2                    bool
crit3                    bool
doubtterr                bool
alternative_id        float64
multiple                 bool
success                  bool
suicide                  bool
targtype1               int64
targtype1_txt          object
targsubtype1          float64
targsubtype1_txt       object
corp1                  object
target1                object
natlty1               float64
natlty1_txt            object
targtype2             float64
targtype2_txt          object
targsubtype2          float64
targsubtype2_txt       object
corp2                  object
target2                object
natlty2               float64
natlty2_txt            object
targtype3             float64
targtype3_txt          object
targsubtype3          float64
targsubtype3_txt       object
corp3     

Iterate through the main dataframe, to create the list of event_ids and the respective 'target_type_ids', 'target_subtype_ids', and 'nationality_ids'.<br />
This list is then used to create the 'event_target' dataframe.

In [43]:
import math
# initialize an empty list
event_targets = []

for row in df.itertuples():
    # append eventid, targtype1, targsubtype1 and natlty1
    event_targets.append(
        {
            'event_id': row.eventid, 
            'target_type_id': int(row.targtype1),
            'target_subtype_id': int(row.targsubtype1),
            'nationality_id': int(row.natlty1)
        }
    )

    # if not all empty, append eventid, targtype2, targsubtype2 and natlty2
    if not (math.isnan(row.targtype2) | math.isnan(row.targsubtype2) | math.isnan(row.natlty2)):
        event_targets.append(
            {
                'event_id': row.eventid, 
                'target_type_id': int(row.targtype2),
                'target_subtype_id': int(row.targsubtype2),
                'nationality_id': int(row.natlty2)
            }
        )

    # if not all empty, append eventid, targtype3, targsubtype3 and natlty3
    if not (math.isnan(row.targtype3) | math.isnan(row.targsubtype3) | math.isnan(row.natlty3)):
        event_targets.append(
            {
                'event_id': row.eventid, 
                'target_type_id': int(row.targtype3),
                'target_subtype_id': int(row.targsubtype3),
                'nationality_id': int(row.natlty3)
            }
        )

df_event_target = pd.DataFrame(event_targets)
df_event_target.head()

event_id  target_type_id  target_subtype_id  nationality_id
0  202001010001              16                 90             141
1  202001010002              16                 90             141
2  202001010003              14                 67             200
3  202001010005              14                 73               4
4  202001010006               8                 49              19

Change the 'event_target' dataframe index to 'event_id', 'target_type_id', 'target_subtype_id', and 'nationality_id', and export to csv.<br />
All columns wwould be used as primary keys in the SQL table.

In [44]:
df_event_target.set_index(['event_id', 'target_type_id', 'target_subtype_id', 'nationality_id']).to_csv('./csv/event_target_data.csv')

Drop all 'targtype*', 'targtype*_txt', 'targsubtype*', 'targsubtype*_txt', 'corp*', 'target*', 'natlty*, and 'natlty*_txt' columns.

In [45]:
df.drop(
    [
        'targtype1', 'targtype1_txt', 'targsubtype1', 'targsubtype1_txt', 
        'corp1', 'target1', 'natlty1', 'natlty1_txt',
        'targtype2', 'targtype2_txt', 'targsubtype2', 'targsubtype2_txt', 
        'corp2', 'target2', 'natlty2', 'natlty2_txt',
        'targtype3', 'targtype3_txt', 'targsubtype3', 'targsubtype3_txt', 
        'corp3', 'target3', 'natlty3', 'natlty3_txt'
    ], 
    axis=1, inplace=True
)

Drop all columns from 'gname' to 'compclaim'

In [46]:
df.drop(
    [
        'gname', 'gsubname', 'gname2', 'gsubname2', 'gname3', 'gsubname3',
        'motive', 'guncertain1', 'guncertain2', 'guncertain3', 'individual',
        'nperps', 'nperpcap', 'claimed', 'claimmode', 'claimmode_txt',
        'claim2', 'claimmode2', 'claimmode2_txt', 'claim3', 'claimmode3', 
        'claimmode3_txt', 'compclaim'
    ], 
    axis=1, inplace=True
)

Iterate through the main dataframe to create the list of 'event_ids' and the respective 'weapon_type_ids', and 'weapon_subtype_ids'.<br />
This list is then used to create the 'event_weapon' dataframe.

In [47]:
import math
# initialize an empty list
event_weapons = []

for row in df.itertuples():
    # append eventid, weaptype1 and weapsubtype1
    event_weapons.append(
        {
            'event_id': row.eventid, 
            'weapon_type_id': int(row.weaptype1),
            'weapon_subtype_id': int(row.weapsubtype1)
        }
    )

    # if not all empty, append eventid, weaptype2, and weapsubtype2
    if not (math.isnan(row.weaptype2) | math.isnan(row.weapsubtype2)):
        event_weapons.append(
            {
                'event_id': row.eventid, 
                'weapon_type_id': int(row.weaptype2),
                'weapon_subtype_id': int(row.weapsubtype2)
            }
        )

    # if not all empty, append eventid, weaptype3, and weapsubtype3
    if not (math.isnan(row.weaptype3) | math.isnan(row.weapsubtype3)):
        event_weapons.append(
            {
                'event_id': row.eventid, 
                'weapon_type_id': int(row.weaptype3),
                'weapon_subtype_id': int(row.weapsubtype3)
            }
        )

    # if not all empty, append eventid, weaptype4, and weapsubtype4
    if not (math.isnan(row.weaptype4) | math.isnan(row.weapsubtype4)):
        event_weapons.append(
            {
                'event_id': row.eventid, 
                'weapon_type_id': int(row.weaptype4),
                'weapon_subtype_id': int(row.weapsubtype4)
            }
        )

df_event_weapon = pd.DataFrame(event_weapons)
df_event_weapon.head()

event_id  weapon_type_id  weapon_subtype_id
0  202001010001               8                 18
1  202001010002               6                 16
2  202001010003               6                 15
3  202001010005               6                  8
4  202001010006               6                 17

Change the 'event_weapon' dataframe index to 'event_id', 'weapon_type_id', and 'weapon_subtype_id', and export to csv. <br />
All columns wwould be used as primary keys in the SQL table.

In [48]:
df_event_weapon.set_index(['event_id', 'weapon_type_id', 'weapon_subtype_id']).to_csv('./csv/event_weapon_data.csv')

Drop all 'weaptype*', 'weaptype*_txt', 'weapsubtype*', and 'weapsubtype*_txt' columns.

In [49]:
df.drop(
    [
        'weaptype1', 'weaptype1_txt', 'weapsubtype1', 'weapsubtype1_txt', 
        'weaptype2', 'weaptype2_txt', 'weapsubtype2', 'weapsubtype2_txt',
        'weaptype3', 'weaptype3_txt', 'weapsubtype3', 'weapsubtype3_txt',
        'weaptype4', 'weaptype4_txt', 'weapsubtype4', 'weapsubtype4_txt'
    ], 
    axis=1, inplace=True
)

Drop all columns after 'nkillter'

In [50]:
df.drop(
    [
        'nwound', 'nwoundus', 'nwoundte', 'property', 'propextent', 'propextent_txt',
        'propvalue', 'propcomment', 'ishostkid', 'nhostkid', 'nhostkidus', 'nhours',
        'ndays', 'divert', 'kidhijcountry', 'ransom', 'ransomamt', 'ransomamtus', 
        'ransompaid', 'ransompaidus', 'ransomnote', 'hostkidoutcome', 'hostkidoutcome_txt',
        'nreleased', 'addnotes', 'scite1', 'scite2', 'scite3', 'dbsource', 'INT_LOG', 
        'INT_IDEO', 'INT_MISC', 'INT_ANY', 'related'
    ], 
    axis=1, inplace=True
)

Re-run dataframe structure commands

In [51]:
df.dtypes

eventid             int64
country_id          int64
region_id           int64
provstate          object
city               object
crit1                bool
crit2                bool
crit3                bool
doubtterr            bool
alternative_id    float64
multiple             bool
success              bool
suicide              bool
weapdetail         object
nkill             float64
nkillus           float64
nkillter          float64
dtype: object